WebDriver: Add support for GetAccessibilityProperties command

rahimabdi opened the pull request:
#### aeb115604ca931db618a0182bc3ffeb3d4f8afea
<pre>
WebDriver: Add support for GetAccessibilityProperties command
<a href="https://bugs.webkit.org/show_bug.cgi?id=299508">https://bugs.webkit.org/show_bug.cgi?id=299508</a>
<a href="https://rdar.apple.com/161303091">rdar://161303091</a>

Reviewed by NOBODY (OOPS!).

This PR implements an internal, test-only WebDriver &quot;GetAccessibilityProperties&quot; command (i.e., not exposed through a WebDriver endpoint) to
support WPT testing. For this first implementation, test_driver.get_accessibility_properties(element) will return an accessibility property bag
containing role/label with future expansion for properties such as checked/disabled/pressed/etc.

Also, the return type for GetComputedLabel (WebAutomationSessionProxy IPC message) incorrectly names the first string parameter as &quot;role&quot;
instead of &quot;label&quot; so this patch corrects the tuple to return (label, errorType).

* LayoutTests/resources/testdriver-vendor.js:
(window.test_driver_internal.get_accessibility_properties):
* Source/WebKit/UIProcess/Automation/Automation.json:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::getAccessibilityProperties):
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::WebAutomationSessionProxy::getAccessibilityProperties):
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.h:
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.messages.in:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fc38c76e6aa434caa68fed1a76432ec34f12d2de

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/162231 "Passed style check") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/35707 "Hash fc38c76e for PR 51306 does not build (failure)") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/28823 "Hash fc38c76e for PR 51306 does not build (failure)") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/171139 "Hash fc38c76e for PR 51306 does not build (failure)") | [❌ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/118604 "Hash fc38c76e for PR 51306 does not build (failure)") 
| | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/35811 "Hash fc38c76e for PR 51306 does not build (failure)") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/35708 "Hash fc38c76e for PR 51306 does not build (failure)") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/171139 "Hash fc38c76e for PR 51306 does not build (failure)") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/59/builds/118604 "Hash fc38c76e for PR 51306 does not build (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/165190 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/155/builds/35811 "Hash fc38c76e for PR 51306 does not build (failure)") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/168/builds/28823 "Hash fc38c76e for PR 51306 does not build (failure)") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/171139 "Hash fc38c76e for PR 51306 does not build (failure)") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/155/builds/35811 "Hash fc38c76e for PR 51306 does not build (failure)") | [❌ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/156/builds/35708 "Hash fc38c76e for PR 51306 does not build (failure)") | [❌ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/18860 "Hash fc38c76e for PR 51306 does not build (failure)") | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/155/builds/35811 "Hash fc38c76e for PR 51306 does not build (failure)") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/168/builds/28823 "Hash fc38c76e for PR 51306 does not build (failure)") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/173633 "Hash fc38c76e for PR 51306 does not build (failure)") | 
| | [❌ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/20986 "Failed to build and analyze WebKit") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/168/builds/28823 "Hash fc38c76e for PR 51306 does not build (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/173633 "Hash fc38c76e for PR 51306 does not build (failure)") | 
| | [❌ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/35381 "Hash fc38c76e for PR 51306 does not build (failure)") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/156/builds/35708 "Hash fc38c76e for PR 51306 does not build (failure)") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/173633 "Hash fc38c76e for PR 51306 does not build (failure)") | 
| | [❌ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/35364 "Hash fc38c76e for PR 51306 does not build (failure)") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/168/builds/28823 "Hash fc38c76e for PR 51306 does not build (failure)") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/97578 "Hash fc38c76e for PR 51306 does not build (failure)") | 
| | [❌ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/160/builds/35364 "Hash fc38c76e for PR 51306 does not build (failure)") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/168/builds/28823 "Hash fc38c76e for PR 51306 does not build (failure)") | | 
| | [❌ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/34874 "Hash fc38c76e for PR 51306 does not build (failure)") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/102626 "Failed to build and analyze WebKit") | | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/34375 "Hash fc38c76e for PR 51306 does not build (failure)") | | | 
| | [❌ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/34620 "Hash fc38c76e for PR 51306 does not build (failure)") | | | 
| | [❌ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/34523 "Hash fc38c76e for PR 51306 does not build (failure)") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/78157100bab6b2fb6db051257d0374afa27b1995. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51306

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3512d01972d519598d810ca20bb98681055323bb. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51306

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3b0c134f8063aa6034d6df804a12969b8815a872. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51306

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/97a19f38df4514ec28e3495e052d93e2e187026f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51306

minorninth commented on Source/WebKit/UIProcess/Automation/Automation.json:
```diff
@@ -2,6 +2,59 @@
+                { "name": "busy", "type": "boolean", "optional": true },
+                { "name": "checked", "type": "string", "optional": true },
+                { "name": "colcount", "type": "integer", "optional": true },
+                { "name": "controlsElements", "type": "string", "optional": true },
```
Why "string" as the type for activeDescendantElement, controlsElements, etc?

minorninth commented on Source/WebKit/UIProcess/Automation/Automation.json:
```diff
@@ -2,6 +2,59 @@
+                { "name": "autocomplete", "type": "string", "optional": true },
+                { "name": "busy", "type": "boolean", "optional": true },
+                { "name": "checked", "type": "string", "optional": true },
+                { "name": "colcount", "type": "integer", "optional": true },
```
This may be following a spec, but the capitalization doesn't seem consistent. Here we have colcount which is all lowercase to match aria-colcount - but above we have activeDescendantElement when the aria property is aria-activedescendant

minorninth commented on Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
```diff
@@ -853,6 +853,79 @@ void WebAutomationSessionProxy::getComputedLabel(WebCore::PageIdentifier pageID,
+
+    axNode.label = axObject->computedLabel();
+    axNode.role = axObject->computedRoleString();
+    axNode.checked = axObject->isChecked();
```
Could we only set the attribute if it applies to that element? For this one you could check axObject->supportsCheckedState(). For something that isn't checkable I think it'd be more correct to leave it as empty/nullopt rather than false.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0c586223c2c8337894699f2c2821b01635b2a080. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51306

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/fd58efdf557a7c392329f77a55bddee5d4212902. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51306

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a0582cc618f33994aee6a2820de51e6978cb543a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51306

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/73ef09ce4e86105734a009898b6782950833dac7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51306

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ff109ef9f680ce8de19fcdbc3393d00a07d03faf. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51306

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/194ef8adc2512e3762e3913f678a8cc27d497543. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51306

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/be2d639ef9d201f59cfea818c901d8c0b2ad7dc5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51306

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/fce763fac1795ad640ec3d6170ba7a457f796345. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51306

rahimabdi committed:
WebDriver: Add support for GetAccessibilityProperties command
https://bugs.webkit.org/show_bug.cgi?id=299508
rdar://161303091

Reviewed by NOBODY (OOPS!).

This PR implements an internal, test-only WebDriver "GetAccessibilityProperties" command (i.e., not exposed through a WebDriver endpoint) to
support WPT testing. For this first implementation, test_driver.get_accessibility_properties(element) will return an accessibility property bag
containing role/label with future expansion for properties such as checked/disabled/pressed/etc.

Also, the return type for GetComputedLabel (WebAutomationSessionProxy IPC message) incorrectly names the first string parameter as "role"
instead of "label" so this patch corrects the tuple to return (label, errorType).

* LayoutTests/resources/testdriver-vendor.js:
(window.test_driver_internal.get_accessibility_properties):
* Source/WebKit/UIProcess/Automation/Automation.json:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::getAccessibilityProperties):
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::WebAutomationSessionProxy::getAccessibilityProperties):
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.h:
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.messages.in:

rahimabdi committed:
WebDriver: Add support for GetAccessibilityProperties command
https://bugs.webkit.org/show_bug.cgi?id=299508
rdar://161303091

Reviewed by NOBODY (OOPS!).

This PR implements an internal, test-only WebDriver "GetAccessibilityProperties" command (i.e., not exposed through a WebDriver endpoint) to
support WPT testing. For this first implementation, test_driver.get_accessibility_properties(element) will return an accessibility property bag
containing role/label with future expansion for properties such as checked/disabled/pressed/etc.

Also, the return type for GetComputedLabel (WebAutomationSessionProxy IPC message) incorrectly names the first string parameter as "role"
instead of "label" so this patch corrects the tuple to return (label, errorType).

* LayoutTests/resources/testdriver-vendor.js:
(window.test_driver_internal.get_accessibility_properties):
* Source/WebKit/UIProcess/Automation/Automation.json:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::getAccessibilityProperties):
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::WebAutomationSessionProxy::getAccessibilityProperties):
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.h:
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.messages.in:

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/aeb115604ca931db618a0182bc3ffeb3d4f8afea. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51306

Commit:
Fix style errors
